python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/utilities/device_operators.cuh>
+
+#include <cuda/atomic>
+
+#include <cooperative_groups.h>
+
+#include <limits>
+#include <type_traits>
+
+template <typename T>
+__device__ bool are_all_nans(cooperative_groups::thread_block const& block,
+                             T const* data,
+                             int64_t size)
+{
+  __shared__ bool result;
+
+  if (block.thread_rank() == 0) { result = true; }
+  block.sync();
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    if (not std::isnan(data[idx])) {
+      result = false;
```

Comment:
Per https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared-memory-5-x

> A shared memory request for a warp does not generate a bank conflict between two threads that access any address within the same 32-bit word (even though the two addresses fall in the same bank). In that case, for read accesses, the word is broadcast to the requesting threads and for write accesses, each address is written by only one of the threads (which thread performs the write is undefined).

So the UB is _which_ thread does the write, but since everyone who might write would write the same thing and a bool is <= 32bits, I am not sure the end result invokes UB. But IANALL.

Reply:
>  the UB is which thread does the write

+1 
All threads writing the same value to a given location is safe since which thread is the writer doesn't really matter.

Reply:
Update, it is UB in the memory model, so we should change the code, the passage I quoted is from the hardware part of the programming guide, the relevant part of the abstract machine memory model is https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#memory-fence-functions

> The CUDA programming model assumes a device with a weakly-ordered memory model, that is the order in which a CUDA thread writes data to shared memory, global memory, page-locked host memory, or the memory of a peer device is not necessarily the order in which the data is observed being written by another CUDA or host thread. **It is undefined behavior for two threads to read from or write to the same memory location without synchronization.**

Emphasis mine.

Reply:
> it is UB in the memory model, so we should change the code

Working on it...

Reply:
Just pushed the changes by using atomic operations. Note this may make the `float64` kernels much slower. I've added a `TODO` reminding us of the future improvement by using CG-level vote functions (like cg.all()) or CUB once block size can be determined at build time. Thank you both for the informative discussion!